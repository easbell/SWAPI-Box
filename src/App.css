.App {
  text-align: center;
  
}

.App-header {
  justify-content: center;
  font-size: 2rem;
  color: #fff;
}

.category {
  color: #fff;
  font-size: 1.5rem;
}
