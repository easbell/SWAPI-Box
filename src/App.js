import React, { Component } from 'react';
import './App.css';
import Splash from './components/Splash/Splash';
import FilterControls from './components/FilterControls/FilterControls';
import CardContainer from './components/CardContainer/CardContainer';
import { fetchCalls } from './fetchCalls';

class App extends Component {
  constructor() {
    super();
    this.state = {
      randomFilm: {},
      cardsSelected: [],
      categorySelected: '',
      errorStatus: ''
    }
  }
  
  async componentDidMount() {
    const random = this.getRandomNum()
    try {
      const fetchedData = await fetchCalls(`https://swapi.co/api/films/${random}`)
      const randomFilm = ({openingCrawl: fetchedData.opening_crawl, title: fetchedData.title, releaseDate: fetchedData.release_date})
      this.setState({randomFilm: randomFilm})
    } catch(error) {
      this.setState({
        errorStatus: error.message,
      })
    }
  }

  getRandomNum = () => {
    return Math.floor(Math.random() * (7 - 1) + 1); 
  }
  
  handleSort = (filter) => {
    const url = `https://swapi.co/api/${filter}`
    this.setState({categorySelected: filter.toUpperCase()})
    if(filter === 'people') {
      this.fetchPeople(url);
    } else if (filter === 'planets') {
      this.fetchPlanets(url)
    } else {
      this.fetchVehicles(url)
    }
  }

  fetchVehicles = async (link) => {
    try {
      const fetchedData = await fetchCalls(link)
      const cardsSelected = await this.simplifyVehicles(fetchedData.results)
      this.setState({ cardsSelected })
    } catch(error) {
      this.setState({
        errorStatus: error.message,
      })
    }
  }

  simplifyVehicles = (allInfo) => {
    return allInfo.map(info => {
      return ({Name: info.name, Model: info.model, Class: info.vehicle_class, Passengers: info.passengers})
    })
  }

  fetchPlanets = async (link) => {
    try {
      const fetchedData = await fetchCalls(link)
      const cardsSelected = await this.fetchAdditionalPlanetInfo(fetchedData.results)
      this.setState( {cardsSelected} )
    } catch(error) {
      this.setState({
        errorStatus: error.message,
      })
    }
  }

  fetchAdditionalPlanetInfo = async (allInfo) => {
    try {
      const fetchedPlanetInfo = allInfo.map(async info => {
        let fetchedResidents = await this.fetchResidents(info)
        return ({Name: info.name, Terrain: info.terrain, Population: info.population, Climate: info.climate, Residents: fetchedResidents})
      })
      return Promise.all(fetchedPlanetInfo)
    } catch(error) {
      this.setState({
        errorStatus: error.message,
      })
    }
  }

  fetchResidents = async (planet) => {
    const residents = planet.residents.map(async resident => {
      const residentInfo = await fetchCalls(resident)
      return residentInfo.name
    })
    return Promise.all(residents)
  }

  fetchPeople = async (link) => {
    try {
      const fetchedData = await fetchCalls(link)
      const cardsSelected = await this.fetchAdditionalPeopleInfo(fetchedData.results)
      this.setState({ cardsSelected })
    } catch(error) {
      this.setState({
        errorStatus: error.message,
      })
    }
  }

  fetchAdditionalPeopleInfo = async (other) => {
    const homeworlds = await this.fetchHomeWorld(other)
    const species = await this.fetchSpecies(other)
    const all = homeworlds.concat(species)
    return this.combineInfo(all)
  }
  
  combineInfo = (allInfo) => {
    const species = allInfo.splice(10, allInfo.length-1)
    let allPeople = [];
    allInfo.forEach((person, i) => {
      allPeople.push(Object.assign({}, person, species[i]));
    });
    return allPeople;
  }
  
  fetchHomeWorld = (homeInfo) => {
    return Promise.all(homeInfo.map(async person => {
      const fetchedData = await fetchCalls(person.homeworld)
      return ({Name: person.name, Homeworld: fetchedData.name, Population: fetchedData.population})       
    }))
  }

  fetchSpecies = (speciesInfo) => {
    return Promise.all(speciesInfo.map(async person => {
      const fetchedData = await fetchCalls(person.species)
      return ({Species: fetchedData.name})
    }))
  }

  render() {
    const { randomFilm, cardsSelected, categorySelected } = this.state;
    return (
      <div className="App">
        <header className="App-header">
          <h1>SWApi Box</h1>
        </header>
        <div>
          <Splash 
            randomFilm={randomFilm}
          />
          <FilterControls 
            handleSort={this.handleSort}
          />
          <h3 className='category'>{categorySelected}</h3>
          <CardContainer 
            cardsSelected={cardsSelected}
            categorySelected={categorySelected}
          />
        </div>
      </div>
    );
  }
}

export default App;
