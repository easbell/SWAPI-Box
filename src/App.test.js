import React from 'react';
import App from './App';
// import { fetchPeople } from './App'

describe('App', () => {
  describe('fetchVehicles', () => {
    let mockData;
  
    beforeEach(() => {
      mockData = {
        name: 'Sand Crawler',
        model: 'Digger Crawler',
        class: 'wheeled',
        passengers: '30'
      }

      fetch = jest.fn().mockImplementation(() => Promise.resolve({
        ok: true,
        status: 200,
        json: () => Promise.resolve(mockData),
      }))
    })
  
    it('calls fetchVehicles with the correct url and options', () => {
      const url = 'www.starwars.com'

      fetchVehicles(url)

      expect(fetch).toHaveBeenCalledWith(url);
    });

  })
})