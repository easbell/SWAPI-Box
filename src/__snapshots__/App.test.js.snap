// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    fetchAdditionalPlanetInfo={[MockFunction]}
    fetchPeopleMocku={[MockFunction]}
    simplifyVehicles={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="App-header"
        >
          <h1>
            SWApi Box
          </h1>
        </header>,
        <div>
          <Splash
            randomFilm={Object {}}
          />
          <FilterControls
            handleSort={[Function]}
          />
          <h3
            className="category"
          >
            
          </h3>
          <CardContainer
            cardsSelected={Array []}
            categorySelected=""
          />
        </div>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h1>
            SWApi Box
          </h1>,
          "className": "App-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "SWApi Box",
          },
          "ref": null,
          "rendered": "SWApi Box",
          "type": "h1",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Splash
              randomFilm={Object {}}
            />,
            <FilterControls
              handleSort={[Function]}
            />,
            <h3
              className="category"
            >
              
            </h3>,
            <CardContainer
              cardsSelected={Array []}
              categorySelected=""
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "randomFilm": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleSort": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
              "className": "category",
            },
            "ref": null,
            "rendered": "",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "cardsSelected": Array [],
              "categorySelected": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="App-header"
          >
            <h1>
              SWApi Box
            </h1>
          </header>,
          <div>
            <Splash
              randomFilm={Object {}}
            />
            <FilterControls
              handleSort={[Function]}
            />
            <h3
              className="category"
            >
              
            </h3>
            <CardContainer
              cardsSelected={Array []}
              categorySelected=""
            />
          </div>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              SWApi Box
            </h1>,
            "className": "App-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "SWApi Box",
            },
            "ref": null,
            "rendered": "SWApi Box",
            "type": "h1",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Splash
                randomFilm={Object {}}
              />,
              <FilterControls
                handleSort={[Function]}
              />,
              <h3
                className="category"
              >
                
              </h3>,
              <CardContainer
                cardsSelected={Array []}
                categorySelected=""
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "randomFilm": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleSort": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "className": "category",
              },
              "ref": null,
              "rendered": "",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "cardsSelected": Array [],
                "categorySelected": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
